### DISCUSSION TITLE
ESP32-S3 and PCM3168A for Multichannel 6 outputs DAC?

### DISCUSSION THREAD

**Discussion by thisisflyte** [pschatzmann/arduino-audio-tools#1757] [Q&A] [answered]
I'm looking to trigger 6 separate wav files simultaneously from an SD card (SPI) at 44.1KHz 16 bit sample rate and output them on separate audio channels using the ESP32-S3. I see that it possible with the Teensy but looking for additional functionally WiFi, BT. This is why the ESP32-S3 is best suited for my application.

It should be possible using the ESP32-S3 and the multi channel audio codec by Texas Instruments PCM3168A which supports 6 in / 8 out and the time-division-multiplexed (TDM) format. I've built the circuit but no success in the firmware communication between the ESP and the Codec. Has anyone managed to successfully achieve multichannel audio output on the ESP32-S3 using the PCM3168A? Any tips would be much appreciated.

**Comment by pschatzmann** [comment, marked as answer]
I never had the chance to test DTM. But I added some DRAFT implementation of this codec to my [audio-drivers project](https://github.com/pschatzmann/arduino-audio-driver).

I suggest that you start simple with a sine test case and expect quite some debugging to make things to work.
Just a couple of comments that I noticed when working on the driver:
- 16 bits (is only right aligned) and is only supported for I2S mode
- For TDM you need to use 24 bit (and I guess 32 might also work)
- The S3 supports TDM on both I2S ports
- Driver currently only implements slave mode: so the ESP32 must be in master mode

To start, I suggest to use the I2SStream from the AudioTools and the PCM3168 class directly from the drivers library. This gives you more control (to set the right format) and less complexity:

Example:
```
  auto config = i2sStream.defaultConfig(TX_MODE);
  config.pin_bck = 14;
  config.pin_ws = 15;
  config.pin_data = 22;
  config.bits_per_sample = 24;
  config.sample_rate = 44100;
  config.channels = 6;
  // config.pin_mck = 0; // optional master clock pin
  // config.i2s_format = I2S_STD_FORMAT; // default format
  // config.is_master = true; // default esp32 is master
  // config.port_no = 0;
  config.signal_type = TDM
  i2sStream.begin(config);
```

When you test the [PCM3168](https://github.com/pschatzmann/arduino-audio-driver/blob/main/src/Driver/pcm3168/pcm3168.h) driver and you have communcations problems: run the Wire scanner to check if the default address is correct.
